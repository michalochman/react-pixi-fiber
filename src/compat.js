import React from "react";

export function areReactHooksAvailable() {
  return typeof React.useEffect === "function";
}

export function isNewContextAvailable() {
  return typeof React.createContext === "function";
}
