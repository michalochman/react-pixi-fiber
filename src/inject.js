import invariant from "fbjs/lib/invariant";

export const INJECTED_TYPES = {};

export function injectType(type, behavior) {
  INJECTED_TYPES[type] = behavior;
  return type;
}

export function createInjectedTypeInstance(type, props, internalInstanceHandle, applyDisplayObjectProps) {
  let instance;

  if (type in INJECTED_TYPES) {
    const injectedType = INJECTED_TYPES[type];
    let customDisplayObject;
    if (typeof injectedType === "function") {
      customDisplayObject = injectedType;
    } else if (typeof injectedType.customDisplayObject === "function") {
      customDisplayObject = injectedType.customDisplayObject;
    }

    invariant(customDisplayObject, "Invalid Component injected to ReactPixiFiber: `%s`.", type);

    instance = customDisplayObject(props);

    if (typeof injectedType.customApplyProps === "function") {
      instance._customApplyProps = injectedType.customApplyProps.bind({
        // See: https://github.com/Izzimach/react-pixi/blob/a25196251a13ed9bb116a8576d93e9fceac2a14c/src/ReactPIXI.js#L953
        applyDisplayObjectProps: applyDisplayObjectProps.bind(null, type, instance),
      });
    }
    if (typeof injectedType.customDidAttach === "function") {
      instance._customDidAttach = injectedType.customDidAttach;
    }
    if (typeof injectedType.customWillDetach === "function") {
      instance._customWillDetach = injectedType.customWillDetach;
    }
  }

  return instance;
}

export function isInjectedType(type) {
  return typeof INJECTED_TYPES[type] !== "undefined";
}
