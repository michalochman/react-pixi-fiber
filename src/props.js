import { TYPES } from "./types";
import * as PIXI from "pixi.js";

export const CHILDREN = "children";

// http://pixijs.download/release/docs/PIXI.DisplayObject.html
const displayObjectDefaultProps = {
  alpha: 1,
  angle: 0,
  buttonMode: false,
  cacheAsBitmap: false,
  cursor: "auto",
  interactive: false,
  pivot: 0,
  position: 0,
  renderable: true,
  rotation: 0,
  scale: 1,
  skew: 0,
  visible: true,
  x: 0,
  y: 0,
};

// http://pixijs.download/release/docs/PIXI.Container.html
const containerDefaultProps = {
  ...displayObjectDefaultProps,
  interactiveChildren: true,
};

// http://pixijs.download/release/docs/PIXI.Sprite.html
const spriteDefaultProps = {
  ...containerDefaultProps,
  anchor: 0,
  blendMode: PIXI.BLEND_MODES.NORMAL,
  pluginName: "batch",
  roundPixels: false,
  tint: 0xffffff,
};

// http://pixijs.download/release/docs/PIXI.extras.BitmapText.html
const bitmapTextDefaultProps = {
  ...containerDefaultProps,
  align: "left",
  anchor: 0,
  letterSpacing: 0,
  maxWidth: 0,
  roundPixels: false,
  text: "",
  tint: 0xffffff,
};

// http://pixijs.download/release/docs/PIXI.Graphics.html
const graphicsDefaultProps = {
  ...containerDefaultProps,
  blendMode: PIXI.BLEND_MODES.NORMAL,
  pluginName: "batch",
  tint: 0xffffff,
};

// http://pixijs.download/release/docs/PIXI.particles.ParticleContainer.html
const particleContainerDefaultProps = {
  ...containerDefaultProps,
  autoResize: false,
  batchSize: 16384,
  blendMode: PIXI.BLEND_MODES.NORMAL,
  interactiveChildren: true,
  maxSize: 1500,
  roundPixels: true,
  tint: 0xffffff,
};

// http://pixijs.download/release/docs/PIXI.Text.html
const textDefaultProps = {
  ...spriteDefaultProps,
  resolution: 1,
  text: "",
};

// http://pixijs.download/release/docs/PIXI.extras.TilingSprite.html
const tilingSpriteDefaultProps = {
  ...spriteDefaultProps,
  clampMargin: 0.5,
  uvRespectAnchor: false,
};

export const defaultProps = {
  [TYPES.BITMAP_TEXT]: bitmapTextDefaultProps,
  [TYPES.CONTAINER]: containerDefaultProps,
  [TYPES.GRAPHICS]: graphicsDefaultProps,
  [TYPES.PARTICLE_CONTAINER]: particleContainerDefaultProps,
  [TYPES.SPRITE]: spriteDefaultProps,
  [TYPES.TEXT]: textDefaultProps,
  [TYPES.TILING_SPRITE]: tilingSpriteDefaultProps,
};
