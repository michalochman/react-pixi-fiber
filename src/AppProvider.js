import React from "react";
import PropTypes from "prop-types";
import { isNewContextAvailable } from "./compat";

const propTypes = {
  app: PropTypes.object.isRequired,
  children: PropTypes.node,
};
const childContextTypes = {
  app: PropTypes.object,
};

let AppContext = null;

function createAppProvider() {
  if (isNewContextAvailable()) {
    // New Context API
    if (AppContext === null) {
      AppContext = React.createContext(null);
    }

    class AppProvider extends React.Component {
      render() {
        const { app, children } = this.props;
        return <AppContext.Provider value={app}>{children}</AppContext.Provider>;
      }
    }

    AppProvider.propTypes = propTypes;

    const withApp = WrappedComponent => {
      function WithApp(props) {
        return <AppContext.Consumer>{app => <WrappedComponent {...props} app={app} />}</AppContext.Consumer>;
      }
      WithApp.displayName = `withApp(${WrappedComponent})`;

      return WithApp;
    };

    return { AppProvider, withApp };
  } else {
    // Legacy Context API
    class AppProvider extends React.Component {
      getChildContext() {
        return {
          app: this.props.app,
        };
      }

      render() {
        return this.props.children;
      }
    }

    AppProvider.propTypes = propTypes;
    AppProvider.childContextTypes = childContextTypes;

    const withApp = WrappedComponent => {
      function WithApp(props, context) {
        return <WrappedComponent {...props} app={context.app} />;
      }
      WithApp.displayName = `withApp(${WrappedComponent})`;
      WithApp.contextTypes = childContextTypes;

      return WithApp;
    };

    return { AppProvider, withApp };
  }
}

const { AppProvider, withApp } = createAppProvider();

export { AppContext, AppProvider, withApp };
