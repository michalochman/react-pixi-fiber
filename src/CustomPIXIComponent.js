import invariant from "fbjs/lib/invariant";
import { injectType } from "./inject";

function CustomPIXIComponent(behavior, type) {
  invariant(
    typeof type === "string",
    "Invalid argument `type` of type `%s` supplied to `CustomPIXIComponent`, expected `string`.",
    typeof type
  );

  return injectType(type, behavior);
}

export default CustomPIXIComponent;
