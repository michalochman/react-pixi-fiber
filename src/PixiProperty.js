// Based on: https://github.com/facebook/react/blob/9c77ffb444598c32c8f92c8d79e406959a10445b/packages/react-dom/src/shared/DOMProperty.js
import { isInjectedType } from "./inject";
import { parsePoint } from "./utils";

// A reserved attribute.
// It is handled by React separately and shouldn't be written to the DOM.
export const RESERVED = 0;

// A simple string attribute.
// Attributes that aren't in the whitelist are presumed to have this type.
export const STRING = 1;

// A real boolean attribute.
// When true, it should be present (set either to an empty string or its name).
export const BOOLEAN = 2;

// An attribute that must be numeric or parse as a numeric.
// When falsy, it should be removed.
export const NUMERIC = 3;

// An attribute that must be positive numeric or parse as a positive numeric.
// When falsy, it should be removed.
export const POSITIVE_NUMERIC = 4;

// An attribute that must be vector or parse as a vector.
// When falsy, it should be removed.
export const VECTOR = 5;

// An attribute that must be function
// When falsy, it should be removed.
export const CALLBACK = 6;

export function shouldIgnoreAttribute(type, name, propertyInfo) {
  if (propertyInfo !== null) {
    return propertyInfo.type === RESERVED;
  }
  if (isInjectedType(type)) {
    return false;
  }
  return false;
}

export function shouldRemoveAttributeWithWarning(type, name, value, propertyInfo) {
  if (propertyInfo !== null && propertyInfo.type === RESERVED) {
    return false;
  }
  switch (typeof value) {
    case "boolean":
      if (isInjectedType(type)) {
        return false;
      }
      return propertyInfo === null || !propertyInfo.acceptsBooleans;
    case "function":
      if (isInjectedType(type)) {
        return false;
      }
      return propertyInfo === null || propertyInfo.type !== CALLBACK;
    case "symbol":
      return true;
    default:
      return false;
  }
}

export function shouldRemoveAttribute(type, name, value, propertyInfo) {
  if (typeof value === "undefined") {
    return true;
  }
  if (shouldRemoveAttributeWithWarning(type, name, value, propertyInfo)) {
    return true;
  }
  if (propertyInfo !== null) {
    switch (propertyInfo.type) {
      case CALLBACK:
        return typeof value !== "function";
      case NUMERIC:
        return isNaN(value);
      case POSITIVE_NUMERIC:
        return isNaN(value) || value < 0;
      case VECTOR:
        const vector = parsePoint(value);
        return vector.length === 0 || vector.findIndex(x => isNaN(x)) !== -1;
    }
  }
  return false;
}

export function getPropertyInfo(name) {
  return properties.hasOwnProperty(name) ? properties[name] : null;
}

function PropertyInfoRecord(name, type) {
  this.acceptsBooleans = type === BOOLEAN;
  this.propertyName = name;
  this.type = type;
}

// When adding attributes to this list, be sure to also add them to
// the `possibleStandardNames` module to ensure casing and incorrect
// name warnings.
const properties = {};

// These props are reserved by React. They shouldn't be written to the DOM.
["children", "parent"].forEach(name => {
  properties[name] = new PropertyInfoRecord(name, RESERVED);
});

// A few React string attributes have a different name.
// This is a mapping from React prop names to the attribute names.
// [["className", "class"]].forEach(([name, attributeName]) => {
//   properties[name] = new PropertyInfoRecord(name, STRING);
// });

// let otherProps = [
//   "align",
//   "blendMode",
//   "canvas",
//   "context",
//   "cursor",
//   "filterArea",
//   "filters",
//   "font",
//   "hitArea",
//   "lineWidth",
//   "mask",
//   "name",
//   "pluginName",
//   "shader",
//   "style",
//   "text",
//   "texture",
//   "tileTransform",
//   "transform",
//   "uvTransform",
// ];

// These are HTML boolean attributes.
[
  "autoResize",
  "buttonMode",
  "cacheAsBitmap",
  "interactive",
  "interactiveChildren",
  "isMask",
  "nativeLines",
  "renderable",
  "roundPixels",
  "uvRespectAnchor",
  "visible",
].forEach(name => {
  properties[name] = new PropertyInfoRecord(name, BOOLEAN);
});

// These are HTML attributes that must be positive numbers.
["alpha", "fillAlpha", "height", "lineColor", "maxWidth", "resolution", "tint", "width"].forEach(name => {
  properties[name] = new PropertyInfoRecord(name, POSITIVE_NUMERIC);
});

// These are HTML attributes that must be numbers.
["boundsPadding", "clampMargin", "rotation", "x", "y"].forEach(name => {
  properties[name] = new PropertyInfoRecord(name, NUMERIC);
});

["anchor", "pivot", "position", "scale", "skew", "tilePosition", "tileScale"].forEach(name => {
  properties[name] = new PropertyInfoRecord(name, VECTOR);
});

[
  "added",
  "click",
  "mousedown",
  "mousemove",
  "mouseout",
  "mouseover",
  "mouseup",
  "mouseupoutside",
  "pointercancel",
  "pointerdown",
  "pointermove",
  "pointerout",
  "pointerover",
  "pointertap",
  "pointerup",
  "pointerupoutside",
  "removed",
  "rightclick",
  "rightdown",
  "rightup",
  "rightupoutside",
  "tap",
  "touchcancel",
  "touchend",
  "touchendoutside",
  "touchmove",
  "touchstart",
].forEach(name => {
  properties[name] = new PropertyInfoRecord(name, CALLBACK);
});
