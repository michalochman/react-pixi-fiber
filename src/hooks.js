import React from "react";
import { useContext, useEffect } from "react";
import { AppContext } from "./AppProvider";

export function usePixiApp() {
  return useContext(AppContext);
}

export function usePixiTicker(fn) {
  const { ticker } = usePixiApp();

  useEffect(() => {
    ticker.add(fn);

    return () => {
      ticker.remove(fn);
    };
  }, [fn, ticker]);
}
