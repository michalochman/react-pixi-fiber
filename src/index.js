import CustomPIXIComponent from "./CustomPIXIComponent";
import { AppContext, AppProvider, withApp } from "./AppProvider";
import Stage, { createStageClass } from "./Stage";
import { TYPES } from "./types";
import { usePixiApp, usePixiTicker } from "./hooks";
import { createRender, createUnmount } from "./render";
import { ReactPixiFiberAsPrimaryRenderer, unstable_batchedUpdates } from "./ReactPixiFiber";
import { applyDisplayObjectProps } from "./ReactPixiFiberComponent";

const render = createRender(ReactPixiFiberAsPrimaryRenderer);
const unmount = createUnmount(ReactPixiFiberAsPrimaryRenderer);

/* Public API */

export {
  AppContext,
  AppProvider,
  CustomPIXIComponent,
  Stage,
  applyDisplayObjectProps,
  createStageClass,
  render,
  unmount,
  withApp,
  usePixiApp,
  usePixiTicker,
  unstable_batchedUpdates,
};

export const BitmapText = TYPES.BITMAP_TEXT;
export const Container = TYPES.CONTAINER;
export const Graphics = TYPES.GRAPHICS;
export const ParticleContainer = TYPES.PARTICLE_CONTAINER;
export const Sprite = TYPES.SPRITE;
export const Text = TYPES.TEXT;
export const TilingSprite = TYPES.TILING_SPRITE;
