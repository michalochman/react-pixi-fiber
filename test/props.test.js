import { defaultProps } from "../src/props";

describe("props", () => {
  describe("defaultProps", () => {
    it("should match", () => {
      expect(defaultProps).toMatchSnapshot();
    });
  });
});
