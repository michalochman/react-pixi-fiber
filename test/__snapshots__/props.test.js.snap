// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props defaultProps should match 1`] = `
Object {
  "BitmapText": Object {
    "align": "left",
    "alpha": 1,
    "anchor": 0,
    "angle": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "letterSpacing": 0,
    "maxWidth": 0,
    "pivot": 0,
    "position": 0,
    "renderable": true,
    "rotation": 0,
    "roundPixels": false,
    "scale": 1,
    "skew": 0,
    "text": "",
    "tint": 16777215,
    "visible": true,
    "x": 0,
    "y": 0,
  },
  "Container": Object {
    "alpha": 1,
    "angle": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "pivot": 0,
    "position": 0,
    "renderable": true,
    "rotation": 0,
    "scale": 1,
    "skew": 0,
    "visible": true,
    "x": 0,
    "y": 0,
  },
  "Graphics": Object {
    "alpha": 1,
    "angle": 0,
    "blendMode": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "pivot": 0,
    "pluginName": "batch",
    "position": 0,
    "renderable": true,
    "rotation": 0,
    "scale": 1,
    "skew": 0,
    "tint": 16777215,
    "visible": true,
    "x": 0,
    "y": 0,
  },
  "ParticleContainer": Object {
    "alpha": 1,
    "angle": 0,
    "autoResize": false,
    "batchSize": 16384,
    "blendMode": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "maxSize": 1500,
    "pivot": 0,
    "position": 0,
    "renderable": true,
    "rotation": 0,
    "roundPixels": true,
    "scale": 1,
    "skew": 0,
    "tint": 16777215,
    "visible": true,
    "x": 0,
    "y": 0,
  },
  "Sprite": Object {
    "alpha": 1,
    "anchor": 0,
    "angle": 0,
    "blendMode": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "pivot": 0,
    "pluginName": "batch",
    "position": 0,
    "renderable": true,
    "rotation": 0,
    "roundPixels": false,
    "scale": 1,
    "skew": 0,
    "tint": 16777215,
    "visible": true,
    "x": 0,
    "y": 0,
  },
  "Text": Object {
    "alpha": 1,
    "anchor": 0,
    "angle": 0,
    "blendMode": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "pivot": 0,
    "pluginName": "batch",
    "position": 0,
    "renderable": true,
    "resolution": 1,
    "rotation": 0,
    "roundPixels": false,
    "scale": 1,
    "skew": 0,
    "text": "",
    "tint": 16777215,
    "visible": true,
    "x": 0,
    "y": 0,
  },
  "TilingSprite": Object {
    "alpha": 1,
    "anchor": 0,
    "angle": 0,
    "blendMode": 0,
    "buttonMode": false,
    "cacheAsBitmap": false,
    "clampMargin": 0.5,
    "cursor": "auto",
    "interactive": false,
    "interactiveChildren": true,
    "pivot": 0,
    "pluginName": "batch",
    "position": 0,
    "renderable": true,
    "rotation": 0,
    "roundPixels": false,
    "scale": 1,
    "skew": 0,
    "tint": 16777215,
    "uvRespectAnchor": false,
    "visible": true,
    "x": 0,
    "y": 0,
  },
}
`;
